From fd12c95ccc0c3ea88759b9911db3e90dc25d6cdc Mon Sep 17 00:00:00 2001
From: David Rosca <david.rosca@amd.com>
Date: Wed, 12 Nov 2025 10:13:32 +0100
Subject: [PATCH] radeonsi/vce: Add workaround for unaligned input surface

VCE requires 16x16 aligned input surface. Blit into an internal
scratch surface when input surface is not 16 aligned.

Cc: mesa-stable
Closes: https://gitlab.freedesktop.org/mesa/mesa/-/issues/14270
Reviewed-by: Ruijing Dong <ruijing.dong@amd.com>
Part-of: <https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/38392>
(cherry picked from commit 17c015c2de7f4a8c5a2459e332424030f2e89ecc)
---
 .pick_status.json                         |  2 +-
 src/gallium/drivers/radeonsi/radeon_vce.c | 25 +++++++++++++++++++++++
 src/gallium/drivers/radeonsi/radeon_vce.h |  1 +
 3 files changed, 27 insertions(+), 1 deletion(-)

diff --git a/.pick_status.json b/.pick_status.json
index c9907b3fbb84d..6dad2bfe00306 100644
--- a/.pick_status.json
+++ b/.pick_status.json
@@ -1154,7 +1154,7 @@
         "description": "radeonsi/vce: Add workaround for unaligned input surface",
         "nominated": true,
         "nomination_type": 1,
-        "resolution": 0,
+        "resolution": 1,
         "main_sha": null,
         "because_sha": null,
         "notes": null
diff --git a/src/gallium/drivers/radeonsi/radeon_vce.c b/src/gallium/drivers/radeonsi/radeon_vce.c
index 9d6f2189f0a22..4389e4b73c8ee 100644
--- a/src/gallium/drivers/radeonsi/radeon_vce.c
+++ b/src/gallium/drivers/radeonsi/radeon_vce.c
@@ -891,6 +891,26 @@ static void rvce_begin_frame(struct pipe_video_codec *encoder, struct pipe_video
    enc->pic = *pic;
    get_param(enc, pic);
 
+   /* VCE requires 16x16 aligned input surface. This is usually not an issue
+    * because surfaces allocated by us are always 16 aligned. However if app
+    * imports external buffer it may not be aligned and then we need to blit
+    * into internal surface used as input surface. */
+   if (source->height % 16) {
+      enc->source_copy = enc->base.context->create_video_buffer(enc->base.context, source);
+      assert(enc->source_copy);
+      struct vl_video_buffer *dst = (struct vl_video_buffer *)enc->source_copy;
+      for (unsigned i = 0; i < 2; i++) {
+         struct pipe_box box = {
+            .width = util_format_get_plane_width(source->buffer_format, i, source->width),
+            .height = util_format_get_plane_height(source->buffer_format, i, source->height),
+            .depth = 1,
+         };
+         si_resource_copy_region(enc->base.context, dst->resources[i], 0, 0, 0, 0, vid_buf->resources[i], 0, &box);
+      }
+      enc->base.context->flush(enc->base.context, NULL, 0);
+      vid_buf = (struct vl_video_buffer *)enc->source_copy;
+   }
+
    enc->get_buffer(vid_buf->resources[0], &enc->handle, &enc->luma);
    enc->get_buffer(vid_buf->resources[1], NULL, &enc->chroma);
 
@@ -1035,6 +1055,11 @@ static int rvce_end_frame(struct pipe_video_codec *encoder, struct pipe_video_bu
 
    flush(enc, picture->flush_flags, picture->out_fence);
 
+   if (enc->source_copy) {
+      enc->source_copy->destroy(enc->source_copy);
+      enc->source_copy = NULL;
+   }
+
    return 0;
 }
 
diff --git a/src/gallium/drivers/radeonsi/radeon_vce.h b/src/gallium/drivers/radeonsi/radeon_vce.h
index 70abb7b3ebb38..a2fc55864a3a6 100644
--- a/src/gallium/drivers/radeonsi/radeon_vce.h
+++ b/src/gallium/drivers/radeonsi/radeon_vce.h
@@ -305,6 +305,7 @@ struct rvce_encoder {
 
    rvce_get_buffer get_buffer;
 
+   struct pipe_video_buffer *source_copy;
    struct pb_buffer_lean *handle;
    struct radeon_surf *luma;
    struct radeon_surf *chroma;
-- 
GitLab

